src/transformer.ts

```diff
@@ -103,37 +103,80 @@ export const transformValue = (
     const { done: valueIsEmpty, value: firstKey } = value.keys().next();
     const returnValueDoesntMatter = valueIsEmpty;
     if (returnValueDoesntMatter || isString(firstKey)) {
-      return { value, type: 'map:string' };
+      return { value, type: ['map', 'string'] };
     }
 
     if (isNumber(firstKey)) {
       return {
         value: value,
-        type: 'map:number',
+        type: ['map', 'number'],
       };
     }
 
     if (isBigint(firstKey)) {
       return {
         value: value,
-        type: 'map:bigint',
+        type: ['map', 'bigint'],
       };
     }
 
     if (isBoolean(firstKey)) {
       return {
         value: value,
-        type: 'map:boolean',
+        type: ['map', 'boolean'],
       };
     }
 
     throw new Error('Key type not supported.');
   }
 
+  if (value?.constructor) {
+    const identifier = ClassRegistry.getIdentifier(value.constructor);
+    if (identifier) {
+      return {
+        value: value,
+        type: ['class', identifier],
+      };
+    }
+  }
+
   return undefined;
 };
 
 export const untransformValue = (json: any, type: TypeAnnotation) => {
+  if (Array.isArray(type)) {
+    switch (type[0]) {
+      case 'map': {
+        switch (type[1]) {
+          case 'number':
+            return new Map(
+              Object.entries(json).map(([k, v]) => [Number(k), v])
+            );
+          case 'string':
+            return new Map(Object.entries(json));
+          case 'boolean':
+            return new Map(
+              Object.entries(json).map(([k, v]) => [Boolean(k), v])
+            );
+          case 'bigint':
+            return new Map(
+              Object.entries(json).map(([k, v]) => [BigInt(k), v])
+            );
+        }
+      }
+
+      case 'class': {
+        const clazz = ClassRegistry.getClass(type[1]);
+
+        if (!clazz) {
+          throw new Error('Trying to deserialize unknown class');
+        }
+
+        return Object.assign(Object.create(clazz.prototype), json);
```

Comment:
This is where the magic happens btw. `Object.create(clazz.prototype)` creates a new, empty instance of `clazz`, which [is much faster](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) than changing the prototype of the existing value using `Object.setPrototypeOf`